src/mast/ast.rs

```diff
@@ -0,0 +1,21 @@
+use crate::{
+    backends::Backend,
+    parser::{Expr, ExprKind},
+};
+
+use super::MastCtx;
+
+impl Expr {
+    /// Convert an expression to another expression, with the same span and a regenerated node id.
+    pub fn to_mast<B: Backend>(&self, ctx: &mut MastCtx<B>, kind: &ExprKind) -> Expr {
+        if !ctx.in_generic_func {
+            return self.clone();
```

Comment:
The thing is we don't know whether a node contains child nodes that are within generic functions. So it needs to walk through the nodes via `monomorphize_expr` and cloning the node.

I think the root cause of the issue is the `monomorphize_expr` needs to propagate the types from both generic and non-generic code. The propagation requires the full AST walking. The `ExprMonoInfo` holds the propagated type and `Expr`, which is either a new monomorphized node or a cloned node.

It seems not trivial to change the field `Expr` of `ExprMonoInfo` to `&Expr`. If this can be done, we could just pass the `&Expr` instead of cloning when it is not within generic functions.